pkg/streamingpromql/operators/functions/absent.go

```diff
@@ -0,0 +1,154 @@
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors.
+
+package functions
+
+import (
+	"context"
+	"errors"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// Absent is an operator that implements the absent() function.
+type Absent struct {
+	timeRange                types.QueryTimeRange
+	argExpressions           parser.Expr
+	inner                    types.InstantVectorOperator
+	expressionPosition       posrange.PositionRange
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	presence                 []bool
+}
+
+var _ types.InstantVectorOperator = &Absent{}
+
+// NewAbsent creates a new Absent.
+func NewAbsent(inner types.InstantVectorOperator, innerExpr parser.Expr, timeRange types.QueryTimeRange, expressionPosition posrange.PositionRange, memoryConsumptionTracker *limiting.MemoryConsumptionTracker) *Absent {
+	return &Absent{
+		timeRange:                timeRange,
+		inner:                    inner,
+		argExpressions:           innerExpr,
+		expressionPosition:       expressionPosition,
+		memoryConsumptionTracker: memoryConsumptionTracker,
+	}
+}
+
+func (a *Absent) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	innerMetadata, err := a.inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer types.PutSeriesMetadataSlice(innerMetadata)
+
+	if a.presence == nil {
+		a.presence, err = types.BoolSlicePool.Get(a.timeRange.StepCount, a.memoryConsumptionTracker)
+		if err != nil {
+			return nil, err
+		}
+
+		// Initialize presence slice
+		for range a.timeRange.StepCount {
+			a.presence = append(a.presence, false)
+		}
+	}
+
+	metadata := types.GetSeriesMetadataSlice(1)
+	metadata = append(metadata, types.SeriesMetadata{
+		Labels: createLabelsForAbsentFunction(a.argExpressions),
+	})
+
+	for range innerMetadata {
+		series, err := a.inner.NextSeries(ctx)
+		if err != nil && errors.Is(err, types.EOS) {
+			return metadata, err
+		}
+		defer types.PutInstantVectorSeriesData(series, a.memoryConsumptionTracker)
+
+		for step := range a.timeRange.StepCount {
+			t := a.timeRange.IndexTime(int64(step))
+			for _, s := range series.Floats {
+				if t == s.T {
+					a.presence[step] = true
+				}
+			}
+			for _, s := range series.Histograms {
+				if t == s.T {
+					a.presence[step] = true
+				}
+			}
+		}
+
+	}
+
+	return metadata, nil
+}
+
+func (a *Absent) NextSeries(_ context.Context) (types.InstantVectorSeriesData, error) {
+	output := types.InstantVectorSeriesData{}
+
+	var err error
+	output.Floats, err = types.FPointSlicePool.Get(a.timeRange.StepCount, a.memoryConsumptionTracker)
```

Comment:
[nit] We could defer allocating this until we know we're returning at least one point in the `if` below

Reply:
Added in 12ee13c41834e59c5b71ff6a1e7a0c36c2f154d9